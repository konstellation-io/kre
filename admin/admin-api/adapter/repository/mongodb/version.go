package mongodb

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"

	"github.com/konstellation-io/kre/admin/admin-api/adapter/config"
	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
	"github.com/konstellation-io/kre/admin/admin-api/domain/usecase"
	"github.com/konstellation-io/kre/admin/admin-api/domain/usecase/logging"
)

type VersionRepoMongoDB struct {
	cfg        *config.Config
	logger     logging.Logger
	collection *mongo.Collection
}

func NewVersionRepoMongoDB(
	cfg *config.Config,
	logger logging.Logger,
	client *mongo.Client,
) *VersionRepoMongoDB {
	collection := client.Database(cfg.MongoDB.DBName).Collection("versions")
	return &VersionRepoMongoDB{
		cfg,
		logger,
		collection,
	}
}

func (r *VersionRepoMongoDB) Create(userID string, newVersion *entity.Version) (*entity.Version, error) {
	newVersion.ID = primitive.NewObjectID().Hex()
	newVersion.CreationDate = time.Now().UTC()
	newVersion.CreationAuthor = userID
	newVersion.Status = entity.VersionStatusStopped
	res, err := r.collection.InsertOne(context.Background(), newVersion)
	if err != nil {
		return nil, err
	}

	newVersion.ID = res.InsertedID.(string)
	return newVersion, nil
}

func (r *VersionRepoMongoDB) GetByID(id string) (*entity.Version, error) {
	v := &entity.Version{}
	filter := bson.D{{"_id", id}}

	err := r.collection.FindOne(context.Background(), filter).Decode(v)
	if err == mongo.ErrNoDocuments {
		return v, usecase.ErrVersionNotFound
	}

	return v, err
}

func (r *VersionRepoMongoDB) GetByIDs(ids []string) ([]*entity.Version, []error) {
	ctx, _ := context.WithTimeout(context.Background(), 60*time.Second)
	var versions []*entity.Version
	cur, err := r.collection.Find(ctx, bson.M{"_id": bson.M{"$in": ids}})
	if err != nil {
		return versions, []error{err}
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var v entity.Version
		err = cur.Decode(&v)
		if err != nil {
			return versions, []error{err}
		}
		versions = append(versions, &v)
	}

	return versions, nil
}

func (r *VersionRepoMongoDB) Update(version *entity.Version) error {
	_, err := r.collection.ReplaceOne(context.Background(), bson.M{"_id": version.ID}, version)
	if err != nil {
		return err
	}

	return nil
}

func (r *VersionRepoMongoDB) GetByRuntime(runtimeID string) ([]*entity.Version, error) {
	ctx, _ := context.WithTimeout(context.Background(), 60*time.Second)
	var versions []*entity.Version
	cur, err := r.collection.Find(ctx, bson.M{"runtimeId": runtimeID})
	if err != nil {
		return versions, err
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var v entity.Version
		err = cur.Decode(&v)
		if err != nil {
			return versions, err
		}
		versions = append(versions, &v)
	}

	return versions, nil
}

func (r *VersionRepoMongoDB) SetHasDoc(ctx context.Context, versionID string, hasDoc bool) error {
	result, err := r.collection.UpdateOne(ctx, bson.M{"_id": versionID}, bson.M{"$set": bson.M{"hasDoc": hasDoc}})
	if err != nil {
		return err
	}

	if result.ModifiedCount != 1 {
		return usecase.ErrVersionNotFound
	}

	return nil
}

func (r *VersionRepoMongoDB) SetStatus(ctx context.Context, versionID string, status entity.VersionStatus) error {
	result, err := r.collection.UpdateOne(ctx, bson.M{"_id": versionID}, bson.M{"$set": bson.M{"status": status}})
	if err != nil {
		return err
	}

	if result.ModifiedCount != 1 {
		return usecase.ErrVersionNotFound
	}

	return nil
}
