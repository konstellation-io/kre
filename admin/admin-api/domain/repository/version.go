package repository

//go:generate mockgen -source=${GOFILE} -destination=$PWD/mocks/repo_${GOFILE} -package=mocks

import (
	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
)

type VersionRepo interface {
	Create(userID string, version *entity.Version) (*entity.Version, error)
	GetByID(id string) (*entity.Version, error)
	GetByIDs(ids []string) ([]*entity.Version, []error)
	Update(version *entity.Version) error
	GetByRuntime(runtimeID string) ([]*entity.Version, error)
}
