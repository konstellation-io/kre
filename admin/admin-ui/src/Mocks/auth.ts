import { AccessLevel } from 'Graphql/types/globalTypes';
import { loader } from 'graphql.macro';

const GetUserEmailQuery = loader('Graphql/queries/getMe.graphql');

export const usernameMock = {
  request: {
    query: GetUserEmailQuery
  },
  result: {
    data: {
      me: {
        email: 'user@konstellation.com',
        id: 'someId',
        accessLevel: AccessLevel.ADMIN
      }
    }
  }
};

export const unauthorizedUsernameMock = {
  request: {
    query: GetUserEmailQuery
  },
  result: {
    data: {
      me: {
        email: 'unauthorizedUser@konstellation.com',
        id: 'someId',
        accessLevel: AccessLevel.ADMIN
      }
    }
  },
  error: new Error('Unauthorized user')
};
