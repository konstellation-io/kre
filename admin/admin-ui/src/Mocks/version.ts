import { RuntimeStatus, VersionStatus } from 'Graphql/types/globalTypes';

import { loader } from 'graphql.macro';

const GetConfigurationVariablesQuery = loader(
  'Graphql/queries/getConfigurationVariables.graphql'
);
const GetVersionWorkflowsQuery = loader(
  'Graphql/queries/getVersionWorkflows.graphql'
);
const GetVersionNodeStatusQuery = loader(
  'Graphql/subscriptions/versionNodeStatus.graphql'
);
const NodeStatusSubscription = loader(
  'Graphql/subscriptions/versionNodeStatus.graphql'
);
const PublishVersionMutation = loader(
  'Graphql/mutations/publishVersion.graphql'
);
const StartVersionMutation = loader('Graphql/mutations/startVersion.graphql');
const StopVersionMutation = loader('Graphql/mutations/stopVersion.graphql');
const UnpublishVersionMutation = loader(
  'Graphql/mutations/unpublishVersion.graphql'
);

export const runtime = {
  id: 'runtimeId',
  name: 'runtimeName',
  description: 'Some Description',
  creationAuthor: { email: 'some@user.com' },
  status: RuntimeStatus.STARTED,
  __typename: 'Runtime'
};
export const version = {
  __typename: 'Version',
  id: 'versionId',
  name: 'versionName',
  description: 'versionDescription',
  status: VersionStatus.STARTED,
  creationDate: '2020/01/01',
  creationAuthor: {
    __typename: 'User',
    id: 'userId',
    email: 'user@domain.com'
  },
  publicationDate: '2020/01/01',
  publicationAuthor: {
    __typename: 'User',
    id: 'userId',
    email: 'user@domain.com'
  },
  config: {
    vars: [],
    completed: false
  }
};
export const confVarsMock = {
  request: {
    query: GetConfigurationVariablesQuery
  },
  result: {
    data: {
      version: {
        status: 'STARTED',
        config: {
          completed: true,
          vars: [
            {
              key: 'var2',
              value: 'value1',
              type: 'VARIABLE'
            },
            {
              key: 'var1',
              value: 'value2',
              type: 'VARIABLE'
            },
            {
              key: 'var3',
              value: 'value3',
              type: 'FILE'
            }
          ]
        }
      }
    }
  }
};

export const workflowsMock = {
  request: {
    query: GetVersionWorkflowsQuery
  },
  result: {
    data: {
      version: {
        name: 'version01',
        status: 'STARTED',
        configurationCompleted: true,
        workflows: [
          {
            name: 'workflow01',
            nodes: [
              {
                id: 'node01',
                name: 'one node',
                status: 'STARTED'
              },
              {
                id: 'node02',
                name: 'another node',
                status: 'STARTED'
              }
            ],
            edges: [
              {
                id: 'edge01',
                fromNode: 'node01',
                toNode: 'node02'
              }
            ]
          }
        ]
      }
    }
  }
};

export const errorMorkflowsMock = {
  request: {
    query: GetVersionWorkflowsQuery
  },
  error: 'Some error'
};

export const nodeStatus = {
  request: {
    query: GetVersionNodeStatusQuery
  },
  result: {
    data: {
      versionNodeStatus: [
        {
          date: '2020-01-01',
          nodeId: 'nodeId1',
          status: 'STARTED',
          message: 'some message 1'
        },
        {
          date: '2020-01-01',
          nodeId: 'nodeId2',
          status: 'STARTED',
          message: 'some message 2'
        }
      ]
    }
  }
};

export const nodeStatusMock = {
  request: {
    query: NodeStatusSubscription
  },
  result: {
    data: {
      versionNodeStatus: {
        date: '2020/01/01',
        nodeId: 'node02',
        status: 'STARTED',
        message: 'some message'
      }
    }
  }
};

export const publishVersionMock = {
  request: {
    query: PublishVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const startVersionMock = {
  request: {
    query: StartVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const stopVersionMock = {
  request: {
    query: StopVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const unpublishVersionMock = {
  request: {
    query: UnpublishVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};
