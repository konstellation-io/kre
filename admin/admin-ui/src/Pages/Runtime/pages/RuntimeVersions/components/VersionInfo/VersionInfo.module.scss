@import 'Styles/mixins';
@import 'Styles/variables';
@import 'Styles/funcs';
@import 'Styles/colors';

$color-border: palette(base, 400);

.container {
  display: flex;
  flex-direction: row;
  border-bottom: solid 1px $color-border;
  transition: background-color ease 0.3s;
  color: font-color(light);

  &.even {
    background-color: $bg-color-lighter;
  }

  .creationDate,
  .name,
  .version,
  .statusLine,
  .statusCircle,
  .description {
    transition: color ease 0.3s, border ease 0.3s, background-color ease 0.3s;
  }

  &:hover {
    background-color: $color-black;
    cursor: pointer;

    .creationDate,
    .name,
    .version,
    .description {
      color: font-color(light);
    }

    .statusLine {
      border-bottom-color: $color-ok;
      border-top-color: $color-ok;
    }

    .statusCircle {
      background-color: $color-ok;
    }
  }

  p {
    margin: 0;
  }
}

.col1 {
  padding: 20px 16px 40px 40px;
  flex-grow: 1;
  max-width: 400px;
  min-width: 340px;
  width: 50%;
  display: flex;
  flex-direction: row;

  .labelContainer {
    .label {
      @include font-button;
      color: $bg-color;
      text-align: center;
      line-height: 24px;

      width: 96px;
      height: 24px;
      margin-right: 24px;

      @each $state in $version-statuses {
        &.#{$state} {
          background-color: get-color-prop($state);
        }
      }
    }
  }

  .actionsContainer {
    width: -webkit-fill-available;
  }

  .creation {
    @include font-caption;
    font-weight: 700;
    line-height: 24px;

    display: flex;
    flex-direction: row;
    align-items: center;
    white-space: nowrap;
    margin-left: auto;
    width: fit-content;
    margin-bottom: 20px;
  }

  .creationDate {
    margin-left: 5px;
  }

  .actionButtons {
    width: fit-content;
    margin-left: auto;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;

    > div {
      margin-left: auto;
    }
  }
}

.col2 {
  width: 50%;
  padding: 20px 20px 40px 30px;
  border-left: solid 1px $color-border;

  .name {
    @include font-body;
    font-weight: 600;
    line-height: 17px;

    margin-bottom: 6px;
  }

  .version {
    @include font-caption;
    line-height: 24px;

    margin-bottom: 15px;
  }

  .descriptionTitle,
  .activatedTitle {
    @include font-caption;
    color: font-color(regular);

    margin-bottom: 8px;
  }

  .activatedTitle {
    margin-top: 15px;
  }

  .col2CreatorName,
  .col2CreationDate {
    display: flex;
    flex-direction: row;
    align-items: center;
  }

  .description {
    @include font-small;
  }

  .activatedContainer {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
  }

  .activatedAuthor,
  .activatedDate {
    @include font-caption;
    line-height: 17px;

    margin: 0 8px 0 8px;
  }
}
