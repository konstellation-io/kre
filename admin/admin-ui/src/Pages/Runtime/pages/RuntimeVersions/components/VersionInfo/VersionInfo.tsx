import CalendarIcon from '@material-ui/icons/Today';
import EmailIcon from '@material-ui/icons/Email';
import { GetVersionConfStatus_versions } from 'Graphql/queries/types/GetVersionConfStatus';
import { Link } from 'react-router-dom';
import ROUTE from 'Constants/routes';
import React from 'react';
import TimeIcon from '@material-ui/icons/AccessTime';
import { VersionStatus } from 'Graphql/types/globalTypes';
import { buildRoute } from 'Utils/routes';
import cx from 'classnames';
import { formatDate } from 'Utils/format';
import styles from './VersionInfo.module.scss';
import { useParams } from 'react-router';

type Props = {
  version: GetVersionConfStatus_versions;
};
function VersionInfo({ version }: Props) {
  const { runtimeId } = useParams();

  const isPublishedVersion = version.status === VersionStatus.PUBLISHED;
  const hasErrors = version.status === VersionStatus.ERROR;

  const versionPath = buildRoute.version(
    ROUTE.RUNTIME_VERSION_STATUS,
    runtimeId,
    version.id
  );

  const errors = version.errors.map((error: string, index: number) => (
    <p key={index} className={styles.versionError}>
      {error}
    </p>
  ));
  return (
    <Link to={versionPath}>
      <div className={styles.container} id={`versionInfoElement_${version.id}`}>
        <div className={styles.col1}>
          <div className={styles.labelContainer}>
            <div className={cx(styles.label, styles[version.status])}>
              {version.status}
            </div>
          </div>
          <div className={styles.actionsContainer}>
            <div className={styles.creation}>
              <CalendarIcon className="icon-small" />
              <p className={styles.creationDate}>
                {`CREATED: ${formatDate(new Date(version.creationDate))}`}
              </p>
            </div>
          </div>
        </div>
        <div className={styles.col2}>
          <p className={styles.version}>{`VERSION ${version.name}`}</p>
          {version.description && (
            <p className={styles.descriptionTitle}>DESCRIPTION</p>
          )}
          <p className={styles.description}>{version.description}</p>
          {isPublishedVersion && (
            <>
              <p className={styles.activatedTitle}>PUBLISHED BY</p>
              <div className={styles.activatedContainer}>
                <div className={styles.col2CreatorName}>
                  <EmailIcon className="icon-small" />
                  <span className={styles.activatedAuthor}>
                    {version &&
                      version.publicationAuthor &&
                      version.publicationAuthor.email}
                  </span>
                </div>
                <div className={styles.col2CreationDate}>
                  <TimeIcon className="icon-small" />
                  <span className={styles.activatedDate}>
                    {version.publicationDate}
                  </span>
                </div>
              </div>
            </>
          )}
          {hasErrors && (
            <>
              <p className={styles.descriptionTitle}>ERRORS</p>
              <div>{errors}</div>
            </>
          )}
        </div>
      </div>
    </Link>
  );
}

export default VersionInfo;
