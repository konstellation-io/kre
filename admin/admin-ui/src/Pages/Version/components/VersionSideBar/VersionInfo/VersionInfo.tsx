import { GetVersionConfStatus_versions } from 'Graphql/queries/types/GetVersionConfStatus';
import React from 'react';
import StatusCircle from 'Components/StatusCircle/StatusCircle';
import cx from 'classnames';
import styles from './VersionInfo.module.scss';

type VersionListItemProps = {
  version: GetVersionConfStatus_versions;
};

function VersionInfo({ version }: VersionListItemProps) {
  return (
    <div className={styles.wrapper}>
      <div className={styles.name}>
        <div>{version.name}</div>
      </div>
      <div className={styles.labelContainer}>
        <StatusCircle status={version.status} />
        <div className={cx(styles.label, styles[version.status])}>
          {version.status}
        </div>
      </div>

      <div>
        <div className={styles.desc}>{version.description}</div>
      </div>
    </div>
  );
}

export default VersionInfo;
