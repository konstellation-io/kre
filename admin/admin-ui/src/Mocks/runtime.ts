import { runtime, version } from './version';

import { loader } from 'graphql.macro';

const GetRuntimesQuery = loader('Graphql/queries/getRuntimes.graphql');
const GetRuntimeAndVersionQuery = loader(
  'Graphql/queries/getRuntimeAndVersions.graphql'
);
const CreateRuntimeMutation = loader('Graphql/mutations/createRuntime.graphql');
const RuntimeCreatedSubscription = loader(
  'Graphql/subscriptions/runtimeCreated.graphql'
);

export const dashboardMock = {
  request: {
    query: GetRuntimesQuery
  },
  result: {
    data: {
      runtimes: [
        {
          id: '00001',
          name: 'Some Name',
          description: 'Some Description',
          status: 'STARTED',
          creationDate: '2019-11-28T15:28:01+00:00',
          creationAuthor: { email: 'some@user.com' },
          publishedVersion: {
            status: 'STARTED'
          }
        },
        {
          id: '00002',
          name: 'Some Other Name',
          description: 'Some Other Description',
          status: 'STARTED',
          creationDate: '2019-11-27T15:28:01+00:00',
          creationAuthor: { email: 'some@user.com' },
          publishedVersion: {
            status: 'STARTED'
          }
        },
        {
          id: '00003',
          name: 'Creating runtime',
          description: 'Some Description',
          status: 'CREATING',
          creationDate: '2019-11-27T15:28:01+00:00',
          creationAuthor: { email: 'some@user.com' },
          publishedVersion: {
            status: 'CREATING'
          }
        }
      ]
    }
  }
};

export const dashboardErrorMock = {
  request: {
    query: GetRuntimesQuery
  },
  error: new Error('cannot get runtimes')
};

export const addRuntimeMock = {
  request: {
    query: CreateRuntimeMutation,
    variables: { input: { name: 'New Runtime' } }
  },
  result: {
    data: {
      createRuntime: { name: 'some name' }
    }
  }
};

export const getRuntimeAndVersionsMock = {
  request: {
    query: GetRuntimeAndVersionQuery
  },
  result: {
    data: {
      runtime,
      version
    }
  }
};

export const getRuntimeAndVersionsErrorMock = {
  request: {
    query: GetRuntimeAndVersionQuery
  },
  error: new Error('cannot get runtime and versions')
};

export const runtimeCreatedMock = {
  request: {
    query: RuntimeCreatedSubscription
  },
  result: {
    data: {
      runtimeCreated: { id: 'some id', name: 'some name' }
    }
  }
};
