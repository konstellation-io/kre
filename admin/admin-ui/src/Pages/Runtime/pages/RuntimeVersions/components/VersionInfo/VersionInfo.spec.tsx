import React from 'react';
import VersionInfo from './VersionInfo';
import { shallow } from 'enzyme';
import { version } from 'Mocks/version';

jest.mock('react-router', () => ({
  useParams: jest.fn(() => ({ runtimeId: 'runtimeId' }))
}));

describe('VersionInfo', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<VersionInfo version={version} />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
