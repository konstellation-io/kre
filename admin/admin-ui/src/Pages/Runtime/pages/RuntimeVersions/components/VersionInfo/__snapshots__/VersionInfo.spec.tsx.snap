// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionInfo matches snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <VersionInfo
      version={
        Object {
          "__typename": "Version",
          "config": Object {
            "completed": false,
            "vars": Array [],
          },
          "creationAuthor": Object {
            "__typename": "User",
            "email": "user@domain.com",
            "id": "userId",
          },
          "creationDate": "2020/01/01",
          "description": "versionDescription",
          "errors": Array [],
          "hasDoc": false,
          "id": "versionId",
          "name": "versionName",
          "publicationAuthor": Object {
            "__typename": "User",
            "email": "user@domain.com",
            "id": "userId",
          },
          "publicationDate": "2020/01/01",
          "status": "STARTED",
        }
      }
    >
      <Link
        to="/runtimes//versions/versionId/status"
      >
        <LinkAnchor
          href="/runtimes//versions/versionId/status"
          navigate={[Function]}
        >
          <a
            href="/runtimes//versions/versionId/status"
            onClick={[Function]}
          >
            <div
              className="container"
              id="versionInfoElement_versionId"
            >
              <div
                className="col1"
              >
                <div
                  className="labelContainer"
                >
                  <div
                    className="label STARTED"
                  >
                    STARTED
                  </div>
                </div>
                <div
                  className="actionsContainer"
                >
                  <div
                    className="creation"
                  >
                    <ForwardRef
                      className="icon-small"
                    >
                      <WithStyles(ForwardRef(SvgIcon))
                        className="icon-small"
                      >
                        <ForwardRef(SvgIcon)
                          className="icon-small"
                          classes={
                            Object {
                              "colorAction": "MuiSvgIcon-colorAction",
                              "colorDisabled": "MuiSvgIcon-colorDisabled",
                              "colorError": "MuiSvgIcon-colorError",
                              "colorPrimary": "MuiSvgIcon-colorPrimary",
                              "colorSecondary": "MuiSvgIcon-colorSecondary",
                              "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                              "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                              "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                              "root": "MuiSvgIcon-root",
                            }
                          }
                        >
                          <svg
                            aria-hidden={true}
                            className="MuiSvgIcon-root icon-small"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z"
                            />
                          </svg>
                        </ForwardRef(SvgIcon)>
                      </WithStyles(ForwardRef(SvgIcon))>
                    </ForwardRef>
                    <p
                      className="creationDate"
                    >
                      CREATED: Jan 1, 2020
                    </p>
                  </div>
                </div>
              </div>
              <div
                className="col2"
              >
                <p
                  className="version"
                >
                  VERSION versionName
                </p>
                <p
                  className="descriptionTitle"
                >
                  DESCRIPTION
                </p>
                <p
                  className="description"
                >
                  versionDescription
                </p>
              </div>
            </div>
          </a>
        </LinkAnchor>
      </Link>
    </VersionInfo>
  </Router>
</BrowserRouter>
`;
